### ISSUE TITLE
Mingw 32bit: Unexpected mark stack overflow (8.0.2+)

### ISSUE THREAD

**Issue by Hamayama** [ivmai/bdwgc#260]
Hello.
I'm using Gauche on Windows.
Recently, Gauche updated bdwgc to 8.0.2 and build/test error began to occur
on MSYS2/MinGW-w64 **32bit** environent.
(There is no problem on MSYS2/MinGW-w64 **64bit** environent.)
(Before update, Gauche used bdwgc 7.6.10 and there was no problem.)

The error message was shown in the message box as follows.
  title:   'Fatal error in GC'
  message: 'Unexpected mark stack overflow'

I tried to investigate this problem and found the following three points.

OS: Windows 8.1 (64bit)
DevTool: MSYS2/MinGW-w64 (32bit) (gcc version 7.3.0 (Rev2, Built by MSYS2 project))

1. gctest error (this might not be a real problem ...)
   I downloaded gc-8.0.2.tar.gz and tested it as follows.
   - **testA:**
     ```
     ./autogen.sh
     ./configure --enable-threads=win32 --enable-large-config --disable-gcj-support CPPFLAGS="-DDONT_ADD_BYTE_AT_END "
     make check
     ```
     **result:**
     TOTAL: 16
     PASS:  16
   
   Then, I tried to gctest repeatedly as follows.
   - **testB:**
     ```
     COUNT=0; ERRCNT=0; while :; do ./gctest.exe || ((ERRCNT++)); ((COUNT++)); echo $COUNT $ERRCNT; done
     ```
     **result:**
     TOTAL COUNT: 200
     ERROR COUNT: 4
   
   Then, I tried testA and testB on the following versions and got almost the same results.
   (testA was passed and testB had some errors)
   - gc master 7bfd248 (2019-1-5)
   - gc master e71ab76 (2018-9-5)
   - gc master 0846bf2 (2016-11-28) + libatomic_ops-7.6.8.tar.gz
   - gc-7.6.10.tar.gz + libatomic_ops-7.6.8.tar.gz
   - gc-7.6.8.tar.gz  + libatomic_ops-7.6.8.tar.gz
   - gc-7.6.8.tar.gz  + libatomic_ops-7.6.6.tar.gz
   - gc-7.4.16.tar.gz + libatomic_ops-7.4.12.tar.gz
   
   The testB error log in .libs/gctest.gc.log had various patterns as follows.
   - 'List reversal produced incorrect list - collector is broken'
   - 'Wrong finalization data - collector is broken'
   - 'Clobbered a leaf - collector is broken'
   - 'Lost a node at level 3 - collector is broken'
   - 'Lost a node at level 9 - collector is broken'
   - 'Lost a node at level 11 - collector is broken'
   - 'Lost a node at level 13 - collector is broken'
   
   But, old Gauche using bdwgc 7.6.10 was no prolem.
   So, I think this might not be a real problem ...

2. 'munmap setting'
   I tried 'git bisect start e71ab76 c7837f8' and found the following commit.
   - gc master 1aabce0 (2018-4-14) 'Enable memory unmapping by default'
   
   I confirmed that the error occurs on Gauche since this commit.
   So, I added configure option `--disable-munmap`
   and confirmed that no error occurs on this commit.
   But, the error still occured on gc-8.0.2.tar.gz .

3. 'new api for VDB'
   I tried 'git bisect start e71ab76 ad5116f' and found the following commit.
   - gc master 06009b0 (2018-8-15) 'New API to turn on manual VDB at runtime'
   
   I confirmed that the error occurs on Gauche since this commit
   even though I added configure option `--disable-munmap`.

For now, I think I should add configure option `--disable-munmap`
for Windows 32bit environment but I don't know how to fix `3.` .
Also, I can't use 'make check' to judge whether Gauche works normally ...

**Comment by ivmai** [comment]
> 'munmap setting'
I've created issue #261 about it.

**Comment by ivmai** [comment]
> gc master 06009b0 (2018-8-15) 'New API to turn on manual VDB at runtime'
This commit has at least a data race bug (see FIXME comment in the commit, which is fixed by 0c0e4cd0b).
Could you please recheck 06009b0 in a single-core mode (i.e. setting affinity to execute gctest on a dedicated cpu core)?

**Comment by ivmai** [comment]
> Could you please recheck 06009b0 in a single-core mode (i.e. setting affinity to execute gctest on a dedicated cpu core)?

Better, please check if the issue is reproducible on commit 815dde3 (belonging to release-8_0 branch) with --disable-munmap

**Comment by Hamayama** [comment]
Thank you for comments. I will try them.

**Comment by ivmai** [comment]
> 1. gctest error (this might not be a real problem ...)

I've created #262 about it.

**Comment by Hamayama** [comment]
I tested gc master 06009b0 .

- **testA:**
  ```
  ./autogen.sh
  ./configure --enable-threads=win32 --enable-large-config --disable-gcj-support CPPFLAGS="-DDONT_ADD_BYTE_AT_END "
  make check
  ```
  **result:**
  TOTAL: 15
  PASS:  15

- **testB:**
  ```
  COUNT=0; ERRCNT=0; while :; do ./gctest.exe || ((ERRCNT++)); ((COUNT++)); echo $COUNT $ERRCNT; done
  ```
  **result:**
  TOTAL COUNT: 200
  ERROR COUNT: 8

- **testB+affinity:**
  ```
  COUNT=0; ERRCNT=0; while :; do cmd /c 'start /b /wait /affinity 1 ./gctest.exe & exit ^!errorlevel^!' || ((ERRCNT++)); ((COUNT++)); echo $COUNT $ERRCNT; done
  ```
  **result:**
  TOTAL COUNT: 1000
  ERROR COUNT: 0

- **testC**
  build/test Gauche without `--disable-munmap`
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

- **testD**
  build/test Gauche with `--disable-munmap`
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

Next, I tested gc release-8_0 815dde3 .

- **testA:**
  **result:**
  TOTAL: 15
  PASS:  15

- **testB:**
  **result:**
  TOTAL COUNT: 200
  ERROR COUNT: 2

- **testB+affinity:**
  **result:**
  TOTAL COUNT: 1000
  ERROR COUNT: 0

- **testC**
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

- **testD**
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

**Comment by Hamayama** [comment]
I found that gc master 06009b0 made a incompatible change
for GC_push_all_stack function in mark.c

I reverted GC_push_all_stack function in mark.c as follows.
```
GC_INNER void GC_push_all_stack(ptr_t bottom, ptr_t top)
{
# if defined(THREADS) && defined(MPROTECT_VDB)
    GC_push_all_eager(bottom, top);
# else
#   ifndef NEED_FIXUP_POINTER
      if (GC_all_interior_pointers) {
        GC_push_all(bottom, top);
      } else
#   endif
    /* else */ {
      GC_push_all_eager(bottom, top);
    }
# endif
}
```

Then, I confirmed that **testD** was passed on following versions.

- gc master 06009b0
- gc release-8_0 815dde3
- gc-8.0.2.tar.gz
- gc-7.6.10.tar.gz + libatomic_ops-7.6.8.tar.gz
- gc master 8076a0c (latest master)

For now, the status of this issue is as follows.

- **testA** - OK
- **testB** - NG -> #262
- **testB+affinity** - OK
- **testC** - NG -> #261
- **testD** - OK ( by reverting GC_push_all_stack in mark.c )

**Comment by ivmai** [comment]
And, if the previous fails, please test master + --enable-munmap (or just w/o --disable-munmap) + reverted GC_push_all_stack with a line changed in the revert:
if defined(THREADS) && defined(MPROTECT_VDB)
->
if defined(THREADS) && (defined(MPROTECT_VDB) || defined(GWW_VDB))

**Comment by Hamayama** [comment]
I tested gc master 8076a0c (latest master) .

- **testC** + GC_push_all_stack-reverted + MPROTECT_VDB-OFF
  ( testC is without `--disable-munmap` )
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

- **testD** + GC_push_all_stack-reverted + MPROTECT_VDB-OFF
  ( testD is with `--disable-munmap` )
  **result:**
  Fatal error in GC「Unexpected mark stack overflow」

- **testC** + GC_push_all_stack-reverted-and-changed
  ( testC is without `--disable-munmap` )
  ( if defined(THREADS) && (defined(MPROTECT_VDB) || defined(GWW_VDB)) )
  **result:**
  OK ! (no error) (I repeated 10 times and all OK)

- **testD** + GC_push_all_stack-reverted-and-changed
  ( testD is with `--disable-munmap` )
  ( if defined(THREADS) && (defined(MPROTECT_VDB) || defined(GWW_VDB)) )
  **result:**
  OK ! (no error) (I repeated 10 times and all OK)

This is the first time that **testC** was passed !
(though I don't understand the reason well ...)

**Comment by ivmai** [comment]
So, fixing this issue will also fix #261.
Quick workaround (on top of master) is:
```
diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index a5f81f43..e1e29767 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -1579,6 +1579,11 @@ EXTERN_C_BEGIN
                 /* os_dep.c.                                            */
 #       define MPROTECT_VDB
 #       define GWW_VDB
+#       ifdef GC_THREADS
+#         define PUSH_ALL_STACK_ALWAYS_EAGER
+                        /* TODO: Otherwise mark stack overflow might    */
+                        /* occur, the reason is unclear.                */
+#       endif
 #       define DATAEND  /* not needed */
 #   endif
 #   ifdef MSWINCE
diff --git a/mark.c b/mark.c
index 9cc9d11c..08d099d1 100644
--- a/mark.c
+++ b/mark.c
@@ -1596,7 +1596,7 @@ GC_API void GC_CALL GC_push_all_eager(void *bottom, void *top)
 
 GC_INNER void GC_push_all_stack(ptr_t bottom, ptr_t top)
 {
-#   ifndef NEED_FIXUP_POINTER
+#   if !defined(NEED_FIXUP_POINTER) && !defined(PUSH_ALL_STACK_ALWAYS_EAGER)
       if (GC_all_interior_pointers
 #         if defined(THREADS) && defined(MPROTECT_VDB)
             && !GC_auto_incremental
```

Should work for you (even without --disable-munmap).

But we'd better figure out the root cause and fix it properly.

**Comment by Hamayama** [comment]
I confirmed that Gauche works with the above patch.

https://ci.appveyor.com/project/Hamayama/gauche/builds/21736322

Thank you !

**Comment by ivmai** [comment]
> I confirmed that Gauche works with the above patch.

Please also test the following patch (instead of the above) - if it works for you, I will merge it.

```
diff --git a/mark.c b/mark.c
index 9cc9d11c..718aaf39 100644
--- a/mark.c
+++ b/mark.c
@@ -1601,7 +1601,8 @@ GC_INNER void GC_push_all_stack(ptr_t bottom, ptr_t top)
 #         if defined(THREADS) && defined(MPROTECT_VDB)
             && !GC_auto_incremental
 #         endif
-         ) {
+          && (word)GC_mark_stack_top
+             < (word)(GC_mark_stack_limit - INITIAL_MARK_STACK_SIZE/8)) {
         GC_push_all(bottom, top);
       } else
 #   endif
```

**Comment by Hamayama** [comment]
I confirmed that Gauche works with the above patch.

https://ci.appveyor.com/project/Hamayama/gauche/builds/21798895

I repeated Gauche's `make check` 10 times on my pc
and confirmed no error.

**Comment by ivmai** [comment]
Thank you for testing. I've merged the final patch to master (some time later I'll back-port it to 8.0 branch, at least). The patch fixes #261 as well.
